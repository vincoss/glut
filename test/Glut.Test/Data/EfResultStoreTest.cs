using Glut.Services;
using Microsoft.Data.Sqlite;
using Microsoft.EntityFrameworkCore;
using System;
using System.Linq;
using System.Collections.Generic;
using System.Text;
using System.Threading.Tasks;
using Xunit;
using ApprovalTests.Reporters;
using ApprovalTests;
using Newtonsoft.Json;
using Moq;
using Glut.Interface;

namespace Glut.Data
{
    public class EfResultStoreTest
    {
        [Theory]
        [InlineData("None", 0)]
        [InlineData("Test", 1)]
        public void GetProjectLastRunIdTest(string projectName, int expected)
        {
            // In-memory database only exists while the connection is open
            var connection = new SqliteConnection("DataSource=:memory:");
            connection.Open();

            try
            {
                var options = new DbContextOptionsBuilder<EfDbContext>()
                    .UseSqlite(connection)
                    .Options;

                // Create the schema in the database
                using (var context = new EfDbContext(options))
                {
                    context.Database.EnsureCreated();
                    new SeedData().Initialize(context);
                    var service = new EfResultStore(context, new EnvironmentService());

                    var actual = service.GetProjectLastRunId(projectName);

                    Assert.Equal(expected, actual);
                }
            }
            finally
            {
                connection.Close();
            }
        }

        [Fact]
        [UseReporter(typeof(DiffReporter))]
        public void AddTest()
        {
            // In-memory database only exists while the connection is open
            var connection = new SqliteConnection("DataSource=:memory:");
            connection.Open();

            try
            {
                var options = new DbContextOptionsBuilder<EfDbContext>()
                    .UseSqlite(connection)
                    .Options;

                // Create the schema in the database
                using (var context = new EfDbContext(options))
                {
                    var environment = new Mock<IEnvironment>();
                    environment.Setup(x => x.SystemDateTimeUtc).Returns(new DateTime(2019, 8, 1));
                    environment.Setup(x => x.UserName).Returns("TestUser");

                    context.Database.EnsureCreated();
                    var service = new EfResultStore(context, environment.Object);

                    var projectName = "TestProject";
                    var attributes = new Dictionary<string, string>();
                    attributes.Add(nameof(projectName), projectName);

                    var start = new DateTime(2019, 8, 1);
                    var result = new ThreadResult();
                    result.Add("/", start, start.AddSeconds(1), true, 200, 100, 200, 300, 400, "Headers", new Exception("Test"));

                    service.Add(projectName, -1, attributes, result);

                    var projects = context.Projects.ToArray();
                    var runAttributes = context.RunAttributes.ToArray();
                    var results = context.Results.ToArray();

                    var sb = new StringBuilder();
                    sb.Append(JsonConvert.SerializeObject(projects));
                    sb.AppendLine();
                    sb.Append(JsonConvert.SerializeObject(runAttributes));
                    sb.AppendLine();
                    sb.Append(JsonConvert.SerializeObject(results));

                    Approvals.Verify(sb.ToString());
                }
            }
            finally
            {
                connection.Close();
            }
        }
    }
}
