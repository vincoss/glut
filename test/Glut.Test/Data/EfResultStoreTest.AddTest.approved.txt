[{"GlutResultId":1,"GlutProjectName":"TestProject","GlutProjectRunId":1,"StartDateTimeUtc":"2019-08-01T00:00:00","EndDateTimeUtc":"2019-08-01T00:00:01","RequestUri":"/","IsSuccessStatusCode":true,"StatusCode":200,"HeaderLength":100,"ResponseLength":200,"TotalLegth":300,"RequestSentTicks":300,"ResponseTicks":400,"TotalTicks":700,"ResponseHeaders":"Headers","Exception":"System.Exception: Test","CreatedDateTimeUtc":"2019-08-01T00:00:00","CreatedByUserName":"TestUser"}]